### ISSUE TITLE
Admin.Client.Upgrade Artifact 

### ISSUE THREAD

**Issue by TotalGriffLock** [Velocidex/velociraptor#215]
I'm trying to use the Admin.Client.Upgrade artifact to push a new version to our clients as per the documentation. When I create a Hunt to push to all clients by tag it targets the correct number of clients, but never advances past that point. If I add a new artifact collection to a single client to perform the same job, I get the following error messages in the browser with an HTTP 500 code:

`{"error":"While parsing source query 0: nary] ) }|: 1:231: unexpected \"\u003cEOF\u003e\" (expected \u003cselect\u003e ...)","code":2,"message":"While parsing source query 0: nary] ) }|: 1:231: unexpected \"\u003cEOF\u003e\" (expected \u003cselect\u003e ...)"}
`

Similar is shown in the VelociraptorGUI_error.log file

`{"error":"{\"error\":\"While parsing source query 0: nary] ) }|: 1:231: unexpected \\\"\\u003cEOF\\u003e\\\" (expected \\u003cselect\\u003e ...)\",\"code\":2,\"message\":\"While parsing source query 0: nary] ) }|: 1:231: unexpected \\\"\\u003cEOF\\u003e\\\" (expected \\u003cselect\\u003e ...)\"}","level":"error","method":"POST","msg":"","remote":"myip:51351","status":500,"time":"2020-01-15T16:28:08Z","url":"/api/v1/CollectArtifact","user":"myusername","user-agent":"Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/79.0.3945.117 Safari/537.36"}
`
I've checked the Admin.Client.Upgrade artifact definition in the tool and it is the same as the one in Github.

The JSON that is posted to CollectArtifact when creating the collection is as follows:

`{"client_id":"C.101bb76b968e0a8e","artifacts":["Admin.Client.Upgrade"],"parameters":{"env":[{"key":"clientURL","value":"https://<<<our_url_removed>>>.msi"}]},"timeout":600}: `

Our clients are running 0.36 and our server is 0.38. This hunt is an attempt to push the upgrade to 0.38 via our own generated msi, which has been tested running manually.

**Comment by scudette** [comment]
There is a YAML encoding bug with this artifact. Normally YAML requires a | to preserve line breaks but this looks ugly and is actually very confusing for users who are not familiar with YAML. Velociraptor hacks around this by recognizing the string "- SELECT" and adding a | in there to preserve the query formatting. But in this case a comment prevents this automatic reformatting.... This causes YAML to swallow all the line feeds and what VQL actually sees is a single line that starts with a comment (so it is all ignored). This is why the error is 'unexpected EOF (expected select)' - VQL is looking for a query that starts with SELECT and not an empty line :-). 

https://github.com/Velocidex/velociraptor/blob/5195c3c12c1f6da96a7dccfd5363484b6420c676/artifacts/definitions/Admin/Client/Upgrade.yaml#L20

So you can easily fix this - simply customize the artifact and then either remove the comment or add a | character like this:
```
queries:
      - |
        // Run the installer on the downloaded file.
        SELECT * from foreach(
         row={
            SELECT Content AS Binary
            FROM http_client(url=clientURL, tempfile_extension=".msi")
         },
         query={
            SELECT * from execve(
               argv=["msiexec.exe", "/i", Binary]
            )
         })
```

If you intend to run this as a hunt to mass upgrade all clients then there is another issue we found - what will happen is that 5k machines will slam your server to download the package at the same time! and will definitely flood your uplinks (unless you are hosting the MSI on some really fat CDN!).

So I would also recommend to add a random sleep to the query which will cause the downloads to be staggered somewhat. So add the following query before to stagger over 10 minutes (might need to increase this if you have a lot of clients):

```
queries:
      - LET _ <= SELECT sleep(time=rand(range=600)) FROM scope()
      - SELECT * from foreach(
         row={
            SELECT Content AS Binary
            FROM http_client(url=clientURL, tempfile_extension=".msi")
         },
         query={
            SELECT * from execve(
               argv=["msiexec.exe", "/i", Binary]
            )
         })
```

**Comment by scudette** [comment]
Hmm I just checked and sleep() and rand() were added Dec 5 2019 in #191 but 0.3.6 was released Nov 12 so it probably wont work on 0.3.6 :-(.  

IIRC this is why we added those functions to be able to stagger the downloads for mass updates. 

An option is to schedule updates using the API on the server side instead of a hunt. You can do this using a server side VQL query.

**Comment by scudette** [comment]
Both these issues should be solved by now.
